// @eddeee888/gcg-typescript-resolver-files/src/generateResolverFiles/handleGraphQLInterfaceType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleGraphQLInterfaceType = void 0;
const utils_1 = require("../utils");
const handleGraphQLInterfaceType = ({ fieldFilePath, resolverName, normalizedResolverName, resolversTypeMeta, moduleName, }, { result, config: { resolverGeneration, emitLegacyCommonJSImports } }) => {
    if (!resolverGeneration.interface) {
        return;
    }
    const resolverTypeString = resolversTypeMeta.typeString;
    const variableStatement = `export const ${resolverName}: ${resolverTypeString} = { /* Implement ${resolverName} interface logic here */ };`;
    result.files[fieldFilePath] = {
        __filetype: 'generalResolver',
        content: `
    ${(0, utils_1.printImportLine)({
            isTypeImport: true,
            module: resolversTypeMeta.module,
            moduleType: resolversTypeMeta.moduleType,
            namedImports: [resolversTypeMeta.typeNamedImport],
            emitLegacyCommonJSImports,
        })}
    ${variableStatement}`,
        mainImportIdentifier: resolverName,
        meta: {
            moduleName,
            normalizedResolverName,
            variableStatement,
            resolverTypeString,
        },
    };
};
exports.handleGraphQLInterfaceType = handleGraphQLInterfaceType;
